sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -566,6 +568,21 @@ object TypeCoercion {
     }
   }
 
+  /**
+   * When all inputs in [[Concat]] are binary, coerces an output type to binary
+   */
+  case class ConcatCoercion(conf: SQLConf) extends TypeCoercionRule {
+    override protected def coerceTypes(
+        plan: LogicalPlan): LogicalPlan = plan transformExpressionsUp {
```

Comment:
This is in the big batch of analyzer rules: `Batch("Resolution")`. This rule will be executed without the stop condition. 

cc @cloud-fan Should we do it in `Batch("Finish Analysis")`? Any better idea?

Reply:
we can add a stop condition `if isBinaryMode = false`

Reply:
From other viewpoints, is that a bad idea to drop this rule from `typeCoercionRules` if `spark.sql.function.concatBinaryAsString` is true? This is like;
```
  def typeCoercionRules(conf: SQLConf): List[Rule[LogicalPlan]] = if (check `spark.sql.function.concatBinaryAsString` ) {
  // rules with ConcatCoercion
} else {
  // rules without ConcatCoercion
}
```